### PR TITLE
expression: add check for utf8 charset when decode to avoid invalid character

### LINKED ISSUE #29685: builtin func char() has different result between MySQL when value is illegal UTF8 character
## Bug Report

builtin func char() has different result between MySQL when value is illegal UTF8 character

### 1. Minimal reproduce step (Required)
```
select char(123456 using utf8);
show warnings;
```
### 2. What did you expect to see? (Required)
```
mysql> select char(123456 using utf8);
+-------------------------+
| char(123456 using utf8) |
+-------------------------+
| NULL                    |
+-------------------------+
1 row in set, 2 warnings (0.00 sec)
mysql> show warnings;
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                                                     |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 3719 | 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous. |
| Warning | 1300 | Invalid utf8mb3 character string: 'E240'                                                                                                                                    |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### 3. What did you see instead (Required)
```
mysql> select char(123456 using utf8);
+-------------------------+
| char(123456 using utf8) |
+-------------------------+
| �@                      |
+-------------------------+
1 row in set (0.01 sec)
mysql> show warnings;
Empty set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

```
mysql> SELECT tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                             |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.2.0-alpha-2186-g2bdebedfc
Edition: Community
Git Commit Hash: 2bdebedfc042e340c7826d0ac96f635f3949a8b3
Git Branch: master
UTC Build Time: 2021-11-11 05:47:41
GoVersion: go1.16.6
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### LINKED ISSUE #29735: Builtin func convert() can't deal with GBK charset correctly
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

start tidb-server with 
```
./bin/tidb-server --config=cmd/explaintest/config.tom
```

```
drop table if exists t;
create table t(a binary(10));
insert into t values (0xe240), (0x01e240);
select convert(a using gbk) from t;
```

### 2. What did you expect to see? (Required)

```
mysql> select convert(a using gbk) from t;
+----------------------+
| convert(a using gbk) |
+----------------------+
| 釦                   |
| 釦                  |
+----------------------+
2 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)

```
mysql> select convert(a using gbk) from t;
ERROR 1105 (HY000): encoding: rune not supported by encoding.
```

### 4. What is your TiDB version? (Required)

```
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                           |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.4.0-alpha-97-g5bece67a9
Edition: Community
Git Commit Hash: 5bece67a9bae979020186b9819b1d3e90cff6678
Git Branch: master
UTC Build Time: 2021-11-12 11:19:21
GoVersion: go1.16.6
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #29685, close #29735

Problem Summary: check the binary string when decode them into utf8

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
check the binary string when convert them to utf8
```

### CODE REVIEW

**Comment by Defined2014** [comment]
/cc @xiongjiwei @tangenta @zimulala @wjhuang2016 
PTAL

**Review by xiongjiwei — COMMENTED** [review]
is it affected by sql_mode?

**Comment by Defined2014** [comment]
> is it affected by sql_mode?


Sql_mode = "", the result is below
```
mysql> SET @@sql_mode='';
Query OK, 0 rows affected (0.00 sec)

mysql> select char(0xe240 using utf8);
+-------------------------+
| char(0xe240 using utf8) |
+-------------------------+
|                         |
+-------------------------+
1 row in set, 2 warnings (0.00 sec)
```

sql_mode=default, the result is below
```
mysql> select char(0xe240 using utf8);
+-------------------------+
| char(0xe240 using utf8) |
+-------------------------+
| NULL                    |
+-------------------------+
1 row in set, 2 warnings (0.00 sec)
```

Seems like `sql-mode` only decide return `NULL` or not. @xiongjiwei

File: parser/charset/encoding.go [hunk 1]
```diff
@@ -140,8 +141,40 @@ func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
...
+		}
+		src = src[size:]
+	}
+
+	return resultStr, nil
+}
+
 // Decode convert bytes from a specific charset to utf-8 charset.
 func (e *Encoding) Decode(dest, src []byte) ([]byte, error) {
+	if e.name == "utf-8" {
```
Comment (wjhuang2016) [inline parser/charset/encoding.go:175]: Please use a const variable.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -1160,29 +1159,26 @@ func (b *builtinConvertSig) evalString(row chunk.Row) (string, bool, error) {
...
 		}
-
-		// we should convert target into utf8 internal.
-		exprInternal, _, _ := transform.String(encoding.NewDecoder(), target)
-		return exprInternal, false, nil
+		return target, false, nil
 	}
 	if types.IsBinaryStr(b.tp) {
 		enc := charset.NewEncoding(b.args[0].GetType().Charset)
 		expr, err = enc.EncodeString(expr)
```
Comment (wjhuang2016) [inline expression/builtin_string.go:1179]: Is it ok to return err directly?

Comment (Defined2014) [reply expression/builtin_string.go:1179]: Not sure. PTAL @xiongjiwei

File: parser/charset/encoding.go [hunk 3]
```diff
@@ -140,8 +141,40 @@ func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
 	return dest
 }
 
+// validUTF8 checks whether there are illegal utf8 characters in []byte.
+func (e *Encoding) validUTF8(src []byte) ([]byte, error) {
```
Comment (xiongjiwei) [inline parser/charset/encoding.go:145]: there is `utf8.Valid()`, we can avoid write it by ourself

File: parser/charset/encoding.go [hunk 4]
```diff
@@ -140,8 +141,40 @@ func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
...
+		}
+		src = src[size:]
+	}
+
+	return resultStr, nil
+}
+
 // Decode convert bytes from a specific charset to utf-8 charset.
 func (e *Encoding) Decode(dest, src []byte) ([]byte, error) {
+	if e.name == encodings[CharsetUTF8].name {
```
Comment (xiongjiwei) [inline parser/charset/encoding.go:175]: @tangenta PTAL, Do we need to check the utf8 here?

File: parser/charset/encoding.go (hunk 3 above)
Comment (Defined2014) [reply parser/charset/encoding.go:145]: Implement this func for compatible err msg.

File: parser/charset/encoding.go (hunk 4 above)
Comment (Defined2014) [reply parser/charset/encoding.go:175]: The reason is https://github.com/pingcap/tidb/issues/29685

File: parser/charset/encoding.go [hunk 5]
```diff
@@ -150,6 +183,9 @@ func (e *Encoding) Decode(dest, src []byte) ([]byte, error) {
 
 // DecodeString convert a string from a specific charset to utf-8 charset.
 func (e *Encoding) DecodeString(src string) (string, error) {
+	if e.name == encodings[CharsetUTF8].name {
```
Comment (bb7133) [inline parser/charset/encoding.go:186]: WIll an extra check cause a performance regression?

Comment (bb7133) [reply parser/charset/encoding.go:186]: DecodeString will called by the lexer.

Comment (Defined2014) [reply parser/charset/encoding.go:186]: I think it will. Maybe #30288 is a better way to do this. I will disscuss with @tangenta later.

Comment (bb7133) [reply parser/charset/encoding.go:186]: @tangenta @Defined2014 Do you still want to go on with this issue, or leave it with #30288?

Comment (Defined2014) [reply parser/charset/encoding.go:186]: Sorry, I will close it.

**Comment by Defined2014** [comment]
It's be covered by #30288.

**Pull request closed without merging**